api/src/notifications/index.js

```diff
@@ -0,0 +1,104 @@
+const prisma = require("../prisma");
+const dayjs = require("dayjs");
+const { sendPushNotification } = require("../services/push-notifications");
+const utc = require("dayjs/plugin/utc");
+dayjs.extend(utc);
+
+const scheduleDefi1Day1 = async (matomoId) => {
+  const type = "DEFI1_DAY1";
+  const user = await prisma.user.findUnique({ where: { matomo_id: matomoId } });
+  if (!user || !user.push_notif_token) return;
+
+  let notif = await prisma.notification.findFirst({
+    where: {
+      userId: user.id,
+      type,
+    },
+  });
+  if (notif) return;
+
+  notif = await prisma.notification.create({
+    data: {
+      user: { connect: { id: user.id } },
+      type,
+      date: dayjs().utc().startOf("day").add(1, "day").toDate(),
+    },
+  });
+};
+
+const removeNotif = async (matomoId, type) => {
+  const user = await prisma.user.findUnique({ where: { matomo_id: matomoId } });
+  if (!user) return;
+
+  await prisma.notification.deleteMany({
+    where: {
+      userId: user.id,
+      type,
+    },
+  });
+};
+
+const notificationsCronJob = async () => {
+  const now = dayjs().utc();
+
+  const reminders = await prisma.reminder.findMany({
+    where: {
+      disabled: false,
+      utcTimeHours: now.hour(),
+      utcTimeMinutes: now.minute(),
+    },
+    include: {
+      user: true,
+      utcDaysOfWeek: true,
+    },
+  });
+
+  const sentNotifications = [];
+  for (const reminder of reminders) {
+    if (!reminder?.user?.push_notif_token) continue;
+
+    const notif = await prisma.notification.findFirst({
```

Comment:
pourquoi rechercher les notifs dans le `for of` des reminders ? les deux sont indépendants me semble-t-il: tu peux avoir un reminder hebdomadaire et pourtant avoir une notif du défi 1 au défi 2, non ?

Reply:
pas complètement indépendants.
si pas de reminder activé, pas de notif
si reminder (aujourd'hui ou mode daily) ET pas de notif, envoyer reminder
si reminder (aujourd'hui ou mode daily) ET notif, envoyer notif
cas particulier : si reminder weekly mais qui ne devait pas tomber aujourd'hui, envoyer notif aujourd'hui quand meme